### PR TITLE
Avoid Crease Defender Clumping

### PR DESCRIPTION
**Pull request by PhilD71**
### Description
We noticed crease defenders tend to "clump" in the corners, often trying to bump each other. This messes with motion planning and is overall good. Changes include:
 - de-mystifying the blocking angle calculation by completely throwing it out and using a simpler version
 - solved clumping
 - stopped goalie from chipping ball in STOP play 
 - fixed bugs in stepAlongPermimeter algorithm
 - Added `isClockwise` to tell if polygons are defined in clockwise form or not

### Testing Done
 - Ai v Ai
 - Unit tests

### Resolved Issues
#2962 #3188

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/proto/parameters.proto [hunk 1]
```diff
@@ -643,4 +643,13 @@ message CreaseDefenderConfig
...
     required double goal_post_offset_chipping = 4
         [default = 0.5, (bounds).min_double_value = 0.0, (bounds).max_double_value = 2.0];
+
+    // determines how much to shift over in the x direction if crease defender
+    // is outside of field boundary
+    required double defender_shift_offset_x = 5 [default = 3.0, (bounds).min_double_value = 0.0, (bounds).max_double_value = 10.0];
+
+    // determines how much to shift over in the y direction if crease defender
+    // is outside the field boundary
+    required double defender_shift_offset_y = 6 [default = 2.0, (bounds).min_double_value = 0.0, (bounds).max_double_value = 10.0];
```
Comment (Andrewyx) [inline src/proto/parameters.proto]: nit: formatting

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 2]
```diff
@@ -11,27 +12,60 @@ std::optional<Point> CreaseDefenderFSM::findBlockThreatPoint(
...
+    std::optional<Point> block_point = findDefenseAreaIntersection(field, ray, inflated_defense_area);
+    double x_shift_offset = crease_defender_config.defender_shift_offset_x();
+    double y_shift_offset = crease_defender_config.defender_shift_offset_y();
+    if (crease_defender_alignment == TbotsProto::CreaseDefenderAlignment::LEFT && block_point.has_value())
+    {
+        // rotate pos 90 deg
+        block_point = stepAlongPerimeter(inflated_defense_area,block_point.value(),-ROBOT_MAX_RADIUS_METERS*2);
+        // Out of field bounds, adjust
+        if (!contains(field.fieldLines(),block_point.value())){
+            /*  In this scenario 0 moves towards x to come back in play.
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:34]: i assume the defense area has a bottom, i.e the block point cannot move more than 1 robot radius out of bounds. cuz otherwise the below moving logic wouldn't work right? maybe add a comment to clarify this

File: src/proto/parameters.proto (hunk 1 above)
Comment (PhilD71) [reply src/proto/parameters.proto]: fixed

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp (hunk 2 above)
Comment (PhilD71) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:34]: fixed

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 3]
```diff
@@ -3,35 +3,72 @@
...
+            *            +-------
+            *          O |
+            *       X    |
+            * -------\ O +------
+            *         \
+            *          0
+            */
+            block_point = block_point.value() + Vector(ROBOT_MAX_RADIUS_METERS*alignment_multiplier_x,ROBOT_MAX_RADIUS_METERS*alignment_multiplier_y);
```
Comment (williamckha) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp]: This works too
```suggestion
            block_point = block_point.value() + ROBOT_MAX_RADIUS_METERS * Vector(alignment_multiplier_x, alignment_multiplier_y);
```

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 4]
```diff
@@ -3,35 +3,72 @@
...
+        // However, it may be out of "playable" boundaries, which isn't useful to us so we move it back inside.
+        if (!contains(field.fieldLines(),block_point.value())){
+            /*  In this scenario 0 moves towards x to come back in play.
+            * -------+
+            *        | O
+            *        |    X
+            * -------+ O / -------
+            *           /
+            *          0
+            */
```
Comment (williamckha) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp]: Nice diagrams!

File: src/proto/parameters.proto [hunk 5]
```diff
@@ -643,4 +643,15 @@ message CreaseDefenderConfig
     // to determine if crease defender chipping is "towards goal"
     required double goal_post_offset_chipping = 4
         [default = 0.5, (bounds).min_double_value = 0.0, (bounds).max_double_value = 2.0];
+
+    // determines how much to shift over in the x direction if crease defender
+    // is outside of field boundary
+    required double defender_shift_offset_x = 5
```
Comment (itsarune) [inline src/proto/parameters.proto]:
```suggestion
    required boundary_robot_radius_shift_offset_x = 5
```

it's a bit unclear what the units of this param is

File: src/proto/parameters.proto [hunk 6]
```diff
@@ -643,4 +643,15 @@ message CreaseDefenderConfig
...
+        [default = 3.0, (bounds).min_double_value = 0.0, (bounds).max_double_value = 10.0];
+
+    // determines how much to shift over in the y direction if crease defender
+    // is outside the field boundary
+    required double defender_shift_offset_y = 6
```
Comment (itsarune) [inline src/proto/parameters.proto]:
```suggestion
    required double boundary_robot_radius_shift_offset_y  = 6
```

File: src/software/ai/motion_constraint/motion_constraint_visitor.cpp [hunk 7]
```diff
@@ -5,8 +5,8 @@
 void MotionConstraintVisitor::visit(const GoalieTactic &tactic)
 {
     current_motion_constraints.erase(TbotsProto::MotionConstraint::FRIENDLY_DEFENSE_AREA);
-    current_motion_constraints.erase(
-        TbotsProto::MotionConstraint::HALF_METER_AROUND_BALL);
+//    current_motion_constraints.erase(
+//        TbotsProto::MotionConstraint::HALF_METER_AROUND_BALL);
```
Comment (itsarune) [inline src/software/ai/motion_constraint/motion_constraint_visitor.cpp]: accidental change, revert

**Review by itsarune — COMMENTED** [review]
looks good, but it's a bit hard to compare without a regression test. could you add one?

**Comment by PhilD71** [comment]
> looks good, but it's a bit hard to compare without a regression test. could you add one?

added tests

File: src/software/geom/polygon.h [hunk 8]
```diff
@@ -50,6 +51,13 @@ class Polygon : public virtual Shape
      */
     double perimeter() const;
 
+    /**
+     * Checks if the segments of the polygon have been provided in clockwise
+     * form
+     * @return
```
Comment (williamckha) [inline src/software/geom/polygon.h]:
```suggestion
     * @return true if the polygon segments are in clockwise order, false otherwise
```

File: src/software/ai/motion_constraint/motion_constraint_visitor.cpp (hunk 7 above)
Comment (PhilD71) [reply src/software/ai/motion_constraint/motion_constraint_visitor.cpp]: pretty sure it needs to go - causes goalie to chip during stop play

File: src/proto/parameters.proto [hunk 9]
```diff
@@ -643,4 +643,12 @@ message CreaseDefenderConfig
     // to determine if crease defender chipping is "towards goal"
     required double goal_post_offset_chipping = 4
         [default = 0.5, (bounds).min_double_value = 0.0, (bounds).max_double_value = 2.0];
+
+    // determines how much to shift over in the x direction if crease defender
+    // is outside of field boundary
+    required double x_shift_meters = 5 [
```
Comment (williamckha) [inline src/proto/parameters.proto]:
```suggestion
    required double out_of_field_boundary_x_shift_meters = 5 [
```

File: src/software/geom/algorithms/step_along_perimeter.cpp [hunk 10]
```diff
@@ -33,15 +33,24 @@ Point stepAlongPerimeter(const Polygon& polygon, const Point& start,
     double perimeter          = polygon.perimeter();
     bool is_counter_clockwise = travel_distance < 0;
     travel_distance           = std::fmod(std::abs(travel_distance), perimeter);
-    if (is_counter_clockwise)
+
+    if (is_counter_clockwise && polygon.isClockwise())
     {
         travel_distance = perimeter - travel_distance;
     }
```
Comment (williamckha) [inline src/software/geom/algorithms/step_along_perimeter.cpp]: Should we account for when the polygon is counter-clockwise and `is_counter_clockwise` is false?

**Review by williamckha — DISMISSED** [review]

File: src/software/geom/algorithms/step_along_perimeter.cpp (hunk 10 above)
Comment (PhilD71) [reply src/software/geom/algorithms/step_along_perimeter.cpp]: we should - good catch!

File: src/software/geom/polygon.h (hunk 8 above)
Comment (PhilD71) [reply src/software/geom/polygon.h]: better

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 11]
```diff
@@ -4,34 +4,81 @@
...
+    double robot_obstacle_inflation_factor, double x_shift_meters)
 {
+    // Look directly at the point
+    // Return the segments that form the path around the crease that the
+    // defenders must follow. It's basically the crease inflated by one robot radius
+    // multiplied by a factor
+    double robot_radius_expansion_amount =
+        ROBOT_MAX_RADIUS_METERS * robot_obstacle_inflation_factor;
+    Rectangle inflated_defense_area =
+        field.friendlyDefenseArea().expand(robot_radius_expansion_amount);
```
Comment (Andrewyx) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp]: This code is duplicated elsewhere, please make a note/minor issue that for post RoboCup (if you would like to do it later) it should be extracted into a function which returns the rectangle. Ideally, this shape should be a constant, but since the field size is technically dynamic a function is required instead.

See: `void CreaseDefenderFSM::blockThreat`

Comment (PhilD71) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp]: fixed

File: src/software/ai/motion_constraint/motion_constraint_visitor.cpp (hunk 7 above)
Comment (PhilD71) [reply src/software/ai/motion_constraint/motion_constraint_visitor.cpp]: nevermind it should probably stay - stop play is a weird edge case we probably wont hit

File: src/proto/parameters.proto [hunk 12]
```diff
@@ -658,4 +658,16 @@ message DefenderStealConfig
     // Max speed of ball that pass defender will try and get possession
     required double max_ball_speed_to_get_m_per_s = 3
         [default = 0.5, (bounds).min_double_value = 0.0, (bounds).max_double_value = 7.0];
+    // The additional buffer length for each side of the goal
+    // to determine if crease defender chipping is "towards goal"
+    required double goal_post_offset_chipping = 4
+        [default = 0.5, (bounds).min_double_value = 0.0, (bounds).max_double_value = 2.0];
```
Comment (Andrewyx) [inline src/proto/parameters.proto:795]: This is already contained within `CreaseDefenderConfig`. Since pass defenders do not have chipping actions, why is this included here?

Comment (PhilD71) [reply src/proto/parameters.proto:795]: oh that must have been added in the git merge - good catch.

File: src/software/geom/polygon.h (hunk 8 above)
Comment (williamckha) [reply src/software/geom/polygon.h]: bump

**Review by williamckha — DISMISSED** [review]
LGTM

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 13]
```diff
@@ -3,34 +3,72 @@
...
+             * -------+
+             *        | O
+             *        |    X
+             * -------+ O / -------
+             *           /
+             *          0
+             */
+            block_point =
+                block_point.value() + Vector(ROBOT_MAX_RADIUS_METERS * 2 + x_shift_meters,
+                                             -ROBOT_MAX_RADIUS_METERS * 2);
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:68]: isn't it possible that the adjusted point is inside the defense area?

**Review by itsarune — COMMENTED** [review]
looks good, have a concern about entering the defense area

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp (hunk 13 above)
Comment (williamckha) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:68]: i don't think this is possible? could you explain how

Comment (PhilD71) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:68]: Yeah, the Y shift always pushes away from defence area

**Review by sauravbanna — APPROVED** [review]

**Review by williamckha — APPROVED** [review]
